File: scripts/ingest_domain_packs.py
#!/usr/bin/env python3
"""
Domain Packs Ingestion Script (Python version)
Parses Domain Packs.md and inserts into Supabase
"""

import os
import re
import json
from supabase import create_client, Client
import openai
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

SUPABASE_URL = os.getenv('VITE_SUPABASE_URL')
SUPABASE_KEY = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
OPENROUTER_KEY = os.getenv('VITE_OPENROUTER_API_KEY')

# Initialize clients
supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)
openai.api_key = OPENROUTER_KEY
openai.api_base = 'https://openrouter.ai/api/v1'

def generate_embedding(text: str):
    """Generate embedding using OpenAI via OpenRouter"""
    response = openai.Embedding.create(
        model='text-embedding-ada-002',
        input=text[:8000]
    )
    return response['data'][0]['embedding']

def parse_domain_packs(file_path: str):
    """Parse the Domain Packs markdown file"""
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()
    
    items = []
    
    # Split by domains
    domain_pattern = r'# \*\*⭐ DOMAIN \d+ — ([^*\n]+)'
    domains = re.split(domain_pattern, content)
    
    for i in range(1, len(domains), 2):
        domain_name = domains[i].strip()
        domain_content = domains[i + 1]
        
        print(f"\n📦 Parsing: {domain_name}")
        
        # Parse principles
        items.extend(parse_principles(domain_name, domain_content))
        
        # Parse mistakes
        items.extend(parse_mistakes(domain_name, domain_content))
        
        # Parse mental models
        items.extend(parse_mental_models(domain_name, domain_content))
        
        # Parse frameworks
        items.extend(parse_frameworks(domain_name, domain_content))
    
    return items

def parse_principles(domain: str, content: str):
    """Extract principles from domain content"""
    items = []
    
    # Match principles section
    match = re.search(r'# \*\*✅.*?PRINCIPLES.*?\n(.*?)(?=# \*\*|$)', content, re.DOTALL | re.IGNORECASE)
    if not match:
        return items
    
    principles_text = match.group(1)
    
    # Extract numbered items
    for line in principles_text.split('\n'):
        m = re.match(r'^\d+\.\s+(.+)$', line.strip())
        if m:
            text = m.group(1).strip()
            # Clean markdown
            text = text.replace('\\>', '>').replace('**', '').replace('*', '')
            if len(text) > 10 and not text.startswith('#'):
                items.append({
                    'domain': domain,
                    'type': 'principle',
                    'name': text[:100],
                    'content': text
                })
    
    print(f"   ✅ Found {len(items)} principles")
    return items

def parse_mistakes(domain: str, content: str):
    """Extract mistakes from domain content"""
    items = []
    
    match = re.search(r'# \*\*❌.*?MISTAKES.*?\n(.*?)(?=# \*\*|$)', content, re.DOTALL | re.IGNORECASE)
    if not match:
        return items
    
    mistakes_text = match.group(1)
    
    for line in mistakes_text.split('\n'):
        m = re.match(r'^\d+\.\s+(.+)$', line.strip())
        if m:
            text = m.group(1).strip()
            text = text.replace('\\>', '>').replace('**', '').replace('*', '')
            if len(text) > 10 and not text.startswith('#'):
                items.append({
                    'domain': domain,
                    'type': 'mistake',
                    'name': text[:100],
                    'content': text
                })
    
    print(f"   ❌ Found {len(items)} mistakes")
    return items

def parse_mental_models(domain: str, content: str):
    """Extract mental models from domain content"""
    items = []
    
    match = re.search(r'# \*\*🧠.*?MENTAL MODELS.*?\n(.*?)(?=# \*\*|$)', content, re.DOTALL | re.IGNORECASE)
    if not match:
        return items
    
    models_text = match.group(1)
    
    # Match pattern: 1. **Name** → Description
    for m in re.finditer(r'\d+\.\s+\*\*(.+?)\*\*\s*[→:\-—]\s*(.+?)(?=\n\d+\.|\n\n|$)', models_text, re.DOTALL):
        name = m.group(1).strip()
        desc = m.group(2).strip()
        if name and desc:
            items.append({
                'domain': domain,
                'type': 'mental_model',
                'name': name,
                'content': f"{name}: {desc}"
            })
    
    print(f"   🧠 Found {len(items)} mental models")
    return items

def parse_frameworks(domain: str, content: str):
    """Extract frameworks from domain content"""
    items = []
    
    match = re.search(r'# \*\*🧩.*?FRAMEWORKS.*?\n(.*?)(?=# \*\*|$)', content, re.DOTALL | re.IGNORECASE)
    if not match:
        return items
    
    frameworks_text = match.group(1)
    
    for line in frameworks_text.split('\n'):
        m = re.match(r'^\d+\.\s+\*\*(.+?)\*\*', line.strip())
        if m:
            name = m.group(1).strip()
            if name:
                items.append({
                    'domain': domain,
                    'type': 'framework',
                    'name': name,
                    'content': name
                })
    
    print(f"   🧩 Found {len(items)} frameworks")
    return items

def get_tags_for_domain(domain: str):
    """Map domain to stage and domain tags"""
    domain_upper = domain.upper()
    
    if 'MARKETING' in domain_upper:
        return {
            'stage_tags': ['mvp', 'first_customers', 'early_revenue', 'pmf', 'growth'],
            'domain_tags': ['Marketing', 'GTM']
        }
    elif 'GTM' in domain_upper or 'GO-TO-MARKET' in domain_upper:
        return {
            'stage_tags': ['first_customers', 'early_revenue', 'pmf', 'seed', 'growth'],
            'domain_tags': ['GTM', 'Strategy']
        }
    elif 'SALES' in domain_upper:
        return {
            'stage_tags': ['first_customers', 'early_revenue', 'pmf', 'seed', 'growth'],
            'domain_tags': ['Sales', 'GTM']
        }
    elif 'TECH' in domain_upper or 'PRODUCT' in domain_upper:
        return {
            'stage_tags': ['idea', 'validation', 'mvp', 'first_customers', 'early_revenue', 'pmf'],
            'domain_tags': ['Product', 'Tech']
        }
    elif 'OPERATIONS' in domain_upper or 'OPS' in domain_upper:
        return {
            'stage_tags': ['first_customers', 'early_revenue', 'pmf', 'seed', 'growth'],
            'domain_tags': ['Ops']
        }
    elif 'LEADERSHIP' in domain_upper or 'HR' in domain_upper:
        return {
            'stage_tags': ['early_revenue', 'pmf', 'seed', 'growth'],
            'domain_tags': ['Leadership', 'HR']
        }
    elif 'FINANCE' in domain_upper:
        return {
            'stage_tags': ['early_revenue', 'pmf', 'seed', 'growth'],
            'domain_tags': ['Finance']
        }
    else:
        return {
            'stage_tags': ['idea', 'validation', 'mvp'],
            'domain_tags': ['Strategy']
        }

def ingest_items(items):
    """Insert items into Supabase"""
    print(f"\n📊 Ingesting {len(items)} items...\n")
    
    inserted = 0
    for i, item in enumerate(items):
        try:
            print(f"[{i+1}/{len(items)}] {item['name'][:60]}...")
            
            # Generate embedding
            embedding = generate_embedding(item['content'])
            
            # Get tags
            tags_info = get_tags_for_domain(item['domain'])
            
            # Determine priority
            priority = 'high' if item['type'] in ['framework', 'principle', 'decision_tree'] else 'medium'
            
            # Insert
            data = {
                'book_title': f"Domain Pack: {item['domain']}",
                'book_author': 'EERA Knowledge Base',
                'book_category': item['domain'],
                'chunk_type': item['type'],
                'chunk_name': item['name'],
                'description': item['content'],
                'components': [item['type'].replace('_', ' ').title()],
                'when_applies': f"Relevant for {', '.join(tags_info['domain_tags'])} at stages: {', '.join(tags_info['stage_tags'])}",
                'tags': [item['domain'].lower(), item['type']],
                'stage_tags': tags_info['stage_tags'],
                'domain_tags': tags_info['domain_tags'],
                'priority': priority,
                'embedding': embedding
            }
            
            result = supabase.table('mentor_knowledge_base').insert(data).execute()
            
            if result.data:
                print(f"   ✅ Inserted")
                inserted += 1
            else:
                print(f"   ❌ Failed")
                
        except Exception as e:
            print(f"   ❌ Error: {str(e)}")
    
    return inserted

def main():
    print("🚀 Domain Packs Ingestion (Python)\n")
    
    file_path = os.path.join(os.getcwd(), 'data', 'mentor-knowledge', 'Domain Packs.md')
    
    if not os.path.exists(file_path):
        print(f"❌ File not found: {file_path}")
        return
    
    # Parse
    items = parse_domain_packs(file_path)
    print(f"\n✅ Parsed {len(items)} total items")
    
    # Ingest
    inserted = ingest_items(items)
    
    print(f"\n{'='*80}")
    print(f"✅ Complete! Inserted {inserted}/{len(items)} items")
    print(f"{'='*80}")

if __name__ == '__main__':
    main()

File: scripts/ingest-domain-packs.ts
/**
 * Domain Packs Ingestion Script
 * 
 * This script:
 * 1. Reads the Domain Packs markdown file
 * 2. Parses each domain (Marketing, Sales, GTM, etc.)
 * 3. Extracts principles, frameworks, mental models, etc.
 * 4. Generates embeddings
 * 5. Inserts into Supabase mentor_knowledge_base table
 * 
 * Usage: npx tsx scripts/ingest-domain-packs.ts
 */

import { createClient } from '@supabase/supabase-js';
import OpenAI from 'openai';
import * as fs from 'fs';
import * as path from 'path';
import dotenv from 'dotenv';

// Load environment variables
dotenv.config();

// Configuration
const SUPABASE_URL = process.env.VITE_SUPABASE_URL;
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;
const OPENROUTER_API_KEY = process.env.VITE_OPENROUTER_API_KEY;

if (!SUPABASE_URL || !SUPABASE_SERVICE_KEY || !OPENROUTER_API_KEY) {
  console.error('❌ Missing environment variables!');
  console.error('Required: VITE_SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY, VITE_OPENROUTER_API_KEY');
  process.exit(1);
}

const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_KEY);
const openai = new OpenAI({ 
  apiKey: OPENROUTER_API_KEY,
  baseURL: 'https://openrouter.ai/api/v1',
});

// Types
interface DomainItem {
  domain: string;
  section: string; // "Core Principles", "Biggest Mistakes", "Mental Models", etc.
  type: string; // "principle", "mistake", "mental_model", "framework", "decision_tree"
  name: string;
  content: string;
  number?: number;
}

// Helper functions
function mapDomainToTags(domain: string): { stage_tags: string[], domain_tags: string[] } {
  const domainMap: Record<string, { stage_tags: string[], domain_tags: string[] }> = {
    'MARKETING': {
      stage_tags: ['mvp', 'first_customers', 'early_revenue', 'pmf', 'growth'],
      domain_tags: ['Marketing', 'GTM']
    },
    'GTM': {
      stage_tags: ['first_customers', 'early_revenue', 'pmf', 'seed', 'growth'],
      domain_tags: ['GTM', 'Strategy']
    },
    'SALES': {
      stage_tags: ['first_customers', 'early_revenue', 'pmf', 'seed', 'growth'],
      domain_tags: ['Sales', 'GTM']
    },
    'TECH': {
      stage_tags: ['idea', 'validation', 'mvp', 'first_customers', 'early_revenue'],
      domain_tags: ['Product', 'Tech']
    },
    'PRODUCT': {
      stage_tags: ['idea', 'validation', 'mvp', 'first_customers', 'early_revenue', 'pmf'],
      domain_tags: ['Product', 'Tech']
    },
    'OPERATIONS': {
      stage_tags: ['first_customers', 'early_revenue', 'pmf', 'seed', 'growth'],
      domain_tags: ['Ops']
    },
    'OPS': {
      stage_tags: ['first_customers', 'early_revenue', 'pmf', 'seed', 'growth'],
      domain_tags: ['Ops']
    },
    'LEADERSHIP': {
      stage_tags: ['first_customers', 'early_revenue', 'pmf', 'seed', 'growth'],
      domain_tags: ['Leadership', 'HR']
    },
    'HR': {
      stage_tags: ['early_revenue', 'pmf', 'seed', 'growth'],
      domain_tags: ['HR', 'Leadership']
    },
    'FINANCE': {
      stage_tags: ['early_revenue', 'pmf', 'seed', 'growth'],
      domain_tags: ['Finance']
    }
  };
  
  // Find matching domain
  for (const [key, tags] of Object.entries(domainMap)) {
    if (domain.toUpperCase().includes(key)) {
      return tags;
    }
  }
  
  return {
    stage_tags: ['idea', 'validation', 'mvp'],
    domain_tags: ['Strategy']
  };
}

async function generateEmbedding(text: string): Promise<number[]> {
  const response = await openai.embeddings.create({
    model: 'text-embedding-ada-002',
    input: text.substring(0, 8000),
  });
  return response.data[0].embedding;
}

function parseDomainPacks(content: string): DomainItem[] {
  const items: DomainItem[] = [];
  
  // Split by domains
  const domainSections = content.split(/# \*\*⭐ DOMAIN \d+ —/);
  
  for (let i = 1; i < domainSections.length; i++) {
    const section = domainSections[i];
    
    // Extract domain name
    const domainMatch = section.match(/^([^*\n]+)/);
    if (!domainMatch) continue;
    
    const domainName = domainMatch[1].trim().replace(/\*\*/g, '');
    console.log(`\n📦 Parsing domain: ${domainName}`);
    
    // Parse different sections
    parsePrinciples(section, domainName, items);
    parseMistakes(section, domainName, items);
    parseMentalModels(section, domainName, items);
    parseFrameworks(section, domainName, items);
    parseDecisionTrees(section, domainName, items);
  }
  
  return items;
}

function parsePrinciples(section: string, domain: string, items: DomainItem[]) {
  // Match both formats: "50 CORE" and "I. 50 CORE"
  const principlesMatch = section.match(/# \*\*✅\s+(?:I\.\s+)?\d+ CORE .+ PRINCIPLES[^#]+([\s\S]*?)(?=# \*\*|$)/i);
  if (!principlesMatch) return;
  
  const principlesText = principlesMatch[1];
  const lines = principlesText.split('\n');
  
  for (const line of lines) {
    const match = line.match(/^\d+\.\s+(.+)$/);
    if (match) {
      const content = match[1].trim();
      // Remove markdown formatting
      const cleanContent = content.replace(/\\/g, '').replace(/\*\*/g, '');
      if (cleanContent && cleanContent.length > 10 && !cleanContent.startsWith('#')) {
        items.push({
          domain,
          section: 'Core Principles',
          type: 'principle',
          name: cleanContent.substring(0, 100),
          content: cleanContent
        });
      }
    }
  }
}

function parseMistakes(section: string, domain: string, items: DomainItem[]) {
  const mistakesMatch = section.match(/# \*\*❌\s+(?:II\.\s+)?\d+ BIGGEST MISTAKES[^#]+([\s\S]*?)(?=# \*\*|$)/i);
  if (!mistakesMatch) return;
  
  const mistakesText = mistakesMatch[1];
  const lines = mistakesText.split('\n');
  
  for (const line of lines) {
    const match = line.match(/^\d+\.\s+(.+)$/);
    if (match) {
      const content = match[1].trim();
      const cleanContent = content.replace(/\\/g, '').replace(/\*\*/g, '');
      if (cleanContent && cleanContent.length > 10 && !cleanContent.startsWith('#')) {
        items.push({
          domain,
          section: 'Biggest Mistakes',
          type: 'mistake',
          name: cleanContent.substring(0, 100),
          content: cleanContent
        });
      }
    }
  }
}

function parseMentalModels(section: string, domain: string, items: DomainItem[]) {
  const modelsMatch = section.match(/# \*\*🧠 \d+ MENTAL MODELS[^#]+([\s\S]*?)(?=# \*\*|$)/i);
  if (!modelsMatch) return;
  
  const modelsText = modelsMatch[1];
  
  // Parse numbered items with descriptions
  const modelMatches = modelsText.matchAll(/^\d+\.\s+\*\*(.+?)\*\*\s*[:\-—]\s*(.+?)(?=\n\d+\.|\n\n|$)/gms);
  
  for (const match of modelMatches) {
    const name = match[1].trim();
    const description = match[2].trim();
    
    if (name && description) {
      items.push({
        domain,
        section: 'Mental Models',
        type: 'mental_model',
        name: name,
        content: `${name}: ${description}`
      });
    }
  }
}

function parseFrameworks(section: string, domain: string, items: DomainItem[]) {
  const frameworksMatch = section.match(/# \*\*🔧 \d+ FRAMEWORKS[^#]+([\s\S]*?)(?=# \*\*|$)/i);
  if (!frameworksMatch) return;
  
  const frameworksText = frameworksMatch[1];
  
  // Parse numbered frameworks with descriptions
  const fwMatches = frameworksText.matchAll(/^\d+\.\s+\*\*(.+?)\*\*\s*[:\-—]\s*(.+?)(?=\n\d+\.|\n\n|$)/gms);
  
  for (const match of fwMatches) {
    const name = match[1].trim();
    const description = match[2].trim();
    
    if (name && description) {
      items.push({
        domain,
        section: 'Frameworks',
        type: 'framework',
        name: name,
        content: `${name}: ${description}`
      });
    }
  }
}

function parseDecisionTrees(section: string, domain: string, items: DomainItem[]) {
  const treesMatch = section.match(/# \*\*🌳 \d+ DECISION TREES[^#]+([\s\S]*?)(?=# \*\*|$)/i);
  if (!treesMatch) return;
  
  const treesText = treesMatch[1];
  
  // Parse decision trees
  const treeMatches = treesText.matchAll(/^\d+\.\s+\*\*(.+?)\*\*\s*[:\-—]?\s*([\s\S]*?)(?=\n\d+\.|\n\n# |$)/gm);
  
  for (const match of treeMatches) {
    const name = match[1].trim();
    const content = match[2].trim();
    
    if (name && content && content.length > 20) {
      items.push({
        domain,
        section: 'Decision Trees',
        type: 'decision_tree',
        name: name,
        content: `${name}: ${content}`
      });
    }
  }
}

async function ingestDomainItems(items: DomainItem[]): Promise<number> {
  console.log(`\n📊 Starting ingestion of ${items.length} items...\n`);
  
  let insertedCount = 0;
  
  for (let i = 0; i < items.length; i++) {
    const item = items[i];
    
    try {
      // Generate embedding
      console.log(`[${i + 1}/${items.length}] Embedding: "${item.name.substring(0, 60)}..."`);
      const embedding = await generateEmbedding(item.content);
      
      // Get tags
      const { stage_tags, domain_tags } = mapDomainToTags(item.domain);
      
      // Determine priority
      const priority = item.type === 'framework' || item.type === 'principle' || item.type === 'decision_tree'
        ? 'high'
        : 'medium';
      
      // Insert into Supabase
      const { error } = await supabase
        .from('mentor_knowledge_base')
        .insert({
          book_title: `Domain Pack: ${item.domain}`,
          book_author: 'EERA Knowledge Base',
          book_category: item.domain,
          chunk_type: item.type,
          chunk_name: item.name,
          description: item.content,
          components: [item.section],
          when_applies: `Relevant for ${domain_tags.join(', ')} at stages: ${stage_tags.join(', ')}`,
          limitations: null,
          tags: [item.domain.toLowerCase(), item.type, item.section.toLowerCase()],
          stage_tags,
          domain_tags,
          priority,
          embedding,
        });
      
      if (error) {
        console.error(`❌ Error inserting "${item.name}":`, error.message);
      } else {
        console.log(`✅ Inserted: "${item.name.substring(0, 60)}"`);
        insertedCount++;
      }
      
      // Rate limit
      await new Promise(resolve => setTimeout(resolve, 20));
      
    } catch (error) {
      console.error(`❌ Error processing "${item.name}":`, error.message);
    }
  }
  
  return insertedCount;
}

async function main() {
  console.log('🚀 Starting Domain Packs Ingestion\n');
  
  // Read the markdown file
  const filePath = path.join(process.cwd(), 'data', 'mentor-knowledge', 'Domain Packs.md');
  
  if (!fs.existsSync(filePath)) {
    console.error(`❌ File not found: ${filePath}`);
    process.exit(1);
  }
  
  console.log(`📖 Reading file: ${filePath}\n`);
  const content = fs.readFileSync(filePath, 'utf-8');
  
  // Parse domain packs
  console.log('🔍 Parsing domain packs...');
  const items = parseDomainPacks(content);
  console.log(`\n✅ Parsed ${items.length} items from domain packs`);
  
  // Show breakdown
  const breakdown = items.reduce((acc, item) => {
    acc[item.domain] = (acc[item.domain] || 0) + 1;
    return acc;
  }, {} as Record<string, number>);
  
  console.log('\n📊 Breakdown by domain:');
  Object.entries(breakdown).forEach(([domain, count]) => {
    console.log(`   - ${domain}: ${count} items`);
  });
  
  // Ingest
  console.log('\n' + '='.repeat(80));
  const startTime = Date.now();
  const inserted = await ingestDomainItems(items);
  const duration = ((Date.now() - startTime) / 1000).toFixed(2);
  
  // Summary
  console.log('\n' + '='.repeat(80));
  console.log('✅ Domain Packs Ingestion Complete!');
  console.log('='.repeat(80));
  console.log(`📊 Summary:`);
  console.log(`   - Items processed: ${items.length}`);
  console.log(`   - Items inserted: ${inserted}`);
  console.log(`   - Duration: ${duration}s`);
  console.log(`   - Average: ${(inserted / parseFloat(duration)).toFixed(2)} items/second`);
  console.log('='.repeat(80));
}

main().catch(console.error);

